src/NServiceBus.Core/Routing/UnicastRoutingTable.cs

```diff
@@ -2,81 +2,51 @@ namespace NServiceBus.Routing
-        /// <param name="route">The route to use for the given message type.</param>
-        /// <exception cref="Exception">Throws an exception when an ambiguous route exists.</exception>
-        public void RouteTo(Type messageType, IUnicastRoute route)
-        {
-            RouteTo(messageType, route, false);
-        }
-
-        /// <summary>
-        /// Adds a static unicast route for a given message type.
-        /// </summary>
-        /// <param name="messageType">The message type to use the route for.</param>
-        /// <param name="route">The route to use for the given message type.</param>
-        /// <param name="overrideExistingRoute">
-        /// Will override an existing route for the message type without throwing an exception
-        /// when set to <code>true</code>.
-        /// </param>
-        /// <exception cref="Exception">
-        /// Throws an exception when an ambiguous route exists and
-        /// <paramref name="overrideExistingRoute" /> is not set to <code>true</code>.
-        /// </exception>
-        public void RouteTo(Type messageType, IUnicastRoute route, bool overrideExistingRoute)
-        {
-            if (!overrideExistingRoute && staticRoutes.ContainsKey(messageType))
-            {
-                throw new Exception($"The static routing table already contains a route for message '{messageType.Name}'. Remove the ambiguous route registrations or override the existing route.");
-            }
-
-            staticRoutes[messageType] = route;
+            return routeTable.TryGetValue(messageType, out unicastRoute) 
+                ? unicastRoute 
+                : null;
         }
 
         /// <summary>
-        /// Configures a dynamically executed fallback rule which is executed when no static route was found for a given message type.
+        /// Adds or replaces a group of routes for a given group key.
         /// </summary>
-        /// <param name="fallbackRoute">The dynamic rule which is invoked to determine the route for a given message type.</param>
-        public void SetFallbackRoute(Func<Type, ContextBag, Task<IUnicastRoute>> fallbackRoute)
+        /// <param name="sourceKey">Key for the route source.</param>
+        /// <param name="entries">Group entries.</param>
+        public void AddOrReplaceRoutes(object sourceKey, IList<RouteTableEntry> entries)
         {
-            if (this.fallbackRoute != null)
+            lock (updateLock)
             {
-                throw new Exception("A custom fallback route has already been configured. Only one fallback route is supported.");
+                routeGroups[sourceKey] = entries;
+                var newRouteTableTemplate = new Dictionary<Type, RouteTableEntry>();
+                foreach (var entry in routeGroups.Values.SelectMany(g => g))
+                {
+                    RouteTableEntry existing;
+                    if (!newRouteTableTemplate.TryGetValue(entry.MessageType, out existing))
```

Comment:
nitpicking, but I think the code would be more readable if we invert the if/else.